### PR TITLE
Fix foreach key evaluation for SimplifyForeachToArrayFilterRector.

### LINKED ISSUE #7592: Incorrect behavior of SimplifyForeachToArrayFilterRector
# Bug Report

| Subject        | Details              |
| :------------- | :--------------------|
| Rector version | last dev-main      |
| Installed as   | composer dependency  |

## Minimal PHP Code Causing Issue

See https://getrector.org/demo/3ae03387-fdc2-469b-9e0c-b0d888a155f1

```php
<?php

class DataFilter
{
    public function filter(array $data): array
    {
        $filteredData = [];
        foreach ($data as $key => $value) {
            if ($key) {
                $filteredData[$key] = $value;
            }
        }

        return $filteredData;
    }
}
```

### Responsible rules

* `SimplifyForeachToArrayFilterRector`

## Expected Behavior

The refactored code miss the definition of `$key` for the callback, as second parameter.

### PR DESCRIPTION
**Pull request by Wohlie**
Fix https://github.com/rectorphp/rector/issues/7592

The issue https://github.com/rectorphp/rector/issues/7570 can be changed to improvement, see expected result.

### CODE REVIEW

File: rules/CodeQuality/Rector/Foreach_/SimplifyForeachToArrayFilterRector.php [hunk 1]
```diff
@@ -119,6 +127,30 @@ private function shouldSkip(Foreach_ $foreach): bool
...
+        }
+
+        $keyVarUsage = $this->betterNodeFinder->find(
+            $ifNode,
+            fn (Node $node): bool => $this->exprUsedInNodeAnalyzer->isUsed($node, $keyVar)
+        );
+
+        $keyVarUsageCount = count($keyVarUsage);
+        if ($keyVarUsageCount === 1) {
+            $parentArrayDimFetch = $keyVarUsage[0]->getAttribute(AttributeKey::PARENT_NODE);
```
Comment (samsonasik) [inline rules/CodeQuality/Rector/Foreach_/SimplifyForeachToArrayFilterRector.php]: `[0]` key usage may be miss re-index, using `current($keyVarUsage)` will be safer.

File: rules/CodeQuality/Rector/Foreach_/SimplifyForeachToArrayFilterRector.php [hunk 2]
```diff
@@ -119,6 +127,30 @@ private function shouldSkip(Foreach_ $foreach): bool
...
+            if (! $parentArrayDimFetch instanceof ArrayDimFetch) {
+                return true;
+            }
+
+            return $parentArrayDimFetch->dim !== $keyVarUsage[0];
```
Comment (samsonasik) [inline rules/CodeQuality/Rector/Foreach_/SimplifyForeachToArrayFilterRector.php]: I tried locally, you can actually use `ReadExprAnalyzer` service:

```suggestion
            return ! $this->readExprAnalyzer->isExprRead(current($keyVarUsage));
```

You need to inject the constructor with `Rector\ReadWrite\NodeAnalyzer\ReadExprAnalyzer` service.

Comment (samsonasik) [reply rules/CodeQuality/Rector/Foreach_/SimplifyForeachToArrayFilterRector.php]: @Wohlie I updated to use `ReadExprAnalyzer` service to your branch https://github.com/rectorphp/rector-src/pull/3100/commits/76fa68122fab6c3e97aa8b0e55534778d97e7555

**Comment by samsonasik** [comment]
Thank you @Wohlie

**Comment by Wohlie** [comment]
Thanks @samsonasik for fixing!

**Pull request merged**